crates/ruff_python_parser/src/parser/statement.rs

```diff
@@ -1564,94 +1590,94 @@ impl<'src> Parser<'src> {
     fn parse_with_item(&mut self, state: WithItemParsingState) -> ParsedWithItem {
         let start = self.node_start();
 
-        let parsed_expr = self.parse_yield_expression_or_else(|p| {
-            p.parse_conditional_expression_or_higher(AllowStarredExpression::Yes)
-        });
         let mut used_ambiguous_lpar = false;
 
-        // While parsing a with item after an ambiguous `(` token, we need to check
-        // for any additional expressions that can be parsed as the above parse function
-        // doesn't do that.
+        // The grammar for the context expression of a with item depends on the state
+        // of with item parsing.
         let context_expr = if state.is_ambiguous_lpar() {
-            match self.current_token_kind() {
-                // Named expressions can come at any position after the ambiguous `(` token.
-                // For example:
-                //
-                // ```python
-                // # Only item
-                // with (item := 10): ...
-                //
-                // # Multiple items
-                // with (item1, item2 := 10): ...
-                // ```
-                TokenKind::ColonEqual => {
-                    let named_expr = self.parse_named_expression(parsed_expr.expr, start);
-
-                    // For example: `with (item := 10 as foo): ...`
-                    if self.at(TokenKind::As) {
-                        self.add_error(
-                            ParseErrorType::OtherError(
-                                "unparenthesized named expression cannot be used here".to_string(),
-                            ),
-                            named_expr.range(),
-                        );
-                    }
```

Comment:
This check is now automatically done when limiting the expression in `parse_parenthesized_with_items`